server/src/compression.rs

```diff
@@ -0,0 +1,473 @@
+    /// The values of the uncompressed segment the compressed segment is being
+    /// build from.
+    uncompressed_values: &'a ValueArray,
+    /// Index of the first data point in `self.uncompressed_timestamps` and
+    /// `self.uncompressed_values` the compressed segment represents.
+    start_index: usize,
+    /// Constant function that currently represents the values in
+    /// `self.uncompressed_values` from `self.start_index` to `self.start_index`
+    /// + `self.pmc_mean.length`.
+    pmc_mean: PMCMean,
+    /// Indicates if `self.pmc_mean` could represent all values in
+    /// `self.uncompressed_values` from `self.start_index` to `current_index` in
+    /// `create_next_compressed_segment()`.
+    pmc_mean_has_fit_all: bool,
+    /// Linear function that represents the values in `self.uncompressed_values`
+    /// from `self.start_index` to `self.start_index` + `self.swing.length`.
+    swing: Swing,
+    /// Indicates if `self.swing` could represent all values in
+    /// `self.uncompressed_values` from `self.start_index` to `current_index` in
+    /// `create_next_compressed_segment()`.
+    swing_has_fit_all: bool,
+    /// Values in `self.uncompressed_values` from `self.start_index` to
+    /// `self.start_index` + `self.gorilla.length` compressed using lossless
+    /// compression.
+    gorilla: Gorilla,
+}
+
+impl<'a> CompressedSegmentBuilder<'a> {
+    fn new(
+        uncompressed_timestamps: &'a TimestampArray,
+        uncompressed_values: &'a ValueArray,
+        error_bound: ErrorBound,
+    ) -> Self {
+        Self {
+            uncompressed_timestamps,
+            uncompressed_values,
+            start_index: 0,
+            pmc_mean: PMCMean::new(error_bound),
+            pmc_mean_has_fit_all: true,
+            swing: Swing::new(error_bound),
+            swing_has_fit_all: true,
+            gorilla: Gorilla::new(),
+        }
+    }
+
+    /// Attempt to update the current models to also represent the `value` of
+    /// the data point collected at `timestamp`.
+    fn try_to_update_models(&mut self, timestamp: Timestamp, value: Value) {
+        self.pmc_mean_has_fit_all = self.pmc_mean_has_fit_all && self.pmc_mean.fit_value(value);
+
+        self.swing_has_fit_all =
+            self.swing_has_fit_all && self.swing.fit_data_point(timestamp, value);
+
+        // Gorilla uses lossless compression and cannot exceed the error bound.
+        if self.gorilla.length < GORILLA_MAXIMUM_LENGTH {
+            self.gorilla.compress_value(value);
+        }
+    }
+
+    /// Return `true` if any of the current models can represent additional
+    /// values, otherwise `false`.
+    fn can_fit_more(&self) -> bool {
+        self.pmc_mean_has_fit_all
+            || self.swing_has_fit_all
+            || self.gorilla.length < GORILLA_MAXIMUM_LENGTH
+    }
+
+    /// Store the model that requires the smallest number of bits per value in
+    /// `compressed_record_batch_builder`. Returns the index of the value in
```

Comment:
Maybe `the index of the first value` instead.

Reply:
Fixed.